arch-design-questions.md

```diff
@@ -0,0 +1,93 @@
+# Client design
+
+## Principles
+
+- There are two top level clients - `ServiceBusReceiverClient` and `ServiceBusSenderClient`.
+- Clients are [single-use](#single-use-clients) - they are created with a receive mode, session ID (if applicable), queue/topic-subscription
+  baked in and cannot be changed.
+- The user can choose between two different methods of message delivery - push (via registered message handlers)
+  or push (via an iterator/reactor pattern or .receiveBatch()-style)
+- Sessions have two different methods for interacting with them:
+  1. [single-use](#single-use-clients)
+  2. [session manager](#sessions-design)
+- Different message types will be used for sending messages vs receiving messages. This is less confusing for
+  the user since we can remove fields that are non-sensical for sending (for instance, `locktoken`).
+
+# Single use clients
+
+Clients are single-use - you cannot change their receive mode or targeted queue/topic/subscription after
+they are created.
+
+This also means that calling .receive() (or registering multiple handlers) isn't a useful activity - the user
+would only be competing against _themselves_ since calling receive() multiple times would just result
+in multiple links, in the same mode, to the same entity, on the same _connection_.
```

Comment:
Calling receive multiple times will not inherently result in multiple links in .NET. There is only one receive link per client. In the ServiceBusProcessorClient, when setting IsSessionEntity = true in the client options, and when using concurrency (MaxConcurrentCalls > 1), each thread will get its own receive link so that we can process multiple sessions simultaneously.

Reply:
Since links are really an implementation detail, should we be calling attention to them at the architectural level?   Granted, they may come into play for context in resource-use scenarios, but it seems like something that is potentially a side-bar here.

Reply:
I removed it. It's probably not important enough and was really just me talking to myself. :)